tdrs-backend/tdpservice/parsers/parse.py

```diff
@@ -61,27 +41,71 @@ def parse_datafile(datafile):
         errors['document'] = ['Section does not match.']
         return errors
 
-    line_errors = parse_datafile_lines(datafile, program_type, section)
+    line_errors = parse_datafile_lines(rawfile, program_type, section)
 
     errors = errors | line_errors
 
     return errors
 
 
-def parse_datafile_lines(datafile, program_type, section):
+def store_record(unsaved_records, record, model):
+    """Store record in dictionary for later processing."""
+    if record:
+        if model not in unsaved_records:
+            unsaved_records.update({model: [record]})
+        else:
+            records_to_save = unsaved_records[model]
+            records_to_save.append(record)
+
+
+def bulk_create_records(unsaved_records):
+    """Bulk create passed in records."""
+    for model, records in unsaved_records.items():
+        model.objects.bulk_create(records)
+
+
+def evaluate_trailer(trailer_count, multiple_trailer_errors, line, errors):
+    """Validate datafile trailer."""
+    if trailer_count > 1 and not multiple_trailer_errors:
+        errors['trailer'] = ['Multiple trailers found.']
+        multiple_trailer_errors = True
+    if trailer_count == 1:
+        record, trailer_is_valid, trailer_errors = schema_defs.trailer.parse_and_validate(line)
+        if not trailer_is_valid:
+            errors['trailer'] = trailer_errors
+
+
+def parse_datafile_lines(rawfile, program_type, section):
     """Parse lines with appropriate schema and return errors."""
     errors = {}
-    rawfile = datafile.file
 
-    rawfile.seek(0)
     line_number = 0
     schema_options = get_schema_options(program_type)
 
+    unsaved_records = {}
+
+    header_count = 0
+    trailer_count = 0
+    prev_sum = 0
+    multiple_trailer_errors = False
+
+    # Note: it is unnecessary to call rawfile.seek(0) again because the generator
+    # automatically starts back at the begining of the file.
     for rawline in rawfile:
         line_number += 1
         line = rawline.decode().strip('\r\n')
 
-        if line.startswith('HEADER') or line.startswith('TRAILER'):
+        header_count += int(line.startswith('HEADER'))
+        trailer_count += int(line.startswith('TRAILER'))
+
+        evaluate_trailer(trailer_count, multiple_trailer_errors, line, errors)
+
+        if header_count > 1:
+            errors.update({'document': ['Multiple headers found.']})
+            return errors
```

Comment:
so, i see what you were going for here - this avoids having to run through the file twice in order to do the "single header/trailer" pre-check, however i think this may result in the unintended effect of allowing every record up until the duplicated header/trailer to be parsed as well, whereas during our initial conversations i recall us wanting to handle it in a different step (so that we would not attempt *any* parsing on a document with multiple headers/trailers). not opposed to this change, just wanted to call out how it changes the parser behavior.

i am curious though, after implementing the `bulk_create` - how much time is actually saved by making the parsing process one loop instead of two? i'd imagine it's negligible compared to the database efficiency change.

Reply:
Ya I definitely wanted feedback on this. On all our existing datafiles it made little difference. However, on the 50MB file it saved ~28 seconds on my machine. That is why I left it in since it was more significant. But the main time killer was definitely the IO so removing this would not be a huge time loss if it is deemed better to die earlier wrt multiple headers/trailers.